x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1138,6 +1145,57 @@ protected In createIn(Expression key, List<Expression> values, ZoneId zoneId) {
         protected Equals createEquals(Expression k, Set<Expression> v, ZoneId finalZoneId) {
             return new Equals(k.source(), k, v.iterator().next(), finalZoneId);
         }
+
+        @Override
+        protected Expression rule(Or or) {
+            Expression e = or;
+            // look only at equals and In
+            List<Expression> exps = splitOr(e);
+
+            Map<Expression, Set<Expression>> found = new LinkedHashMap<>();
+            ZoneId zoneId = null;
+            List<Expression> ors = new LinkedList<>();
+
+            for (Expression exp : exps) {
+                if (exp instanceof Equals eq) {
```

Comment:
Note, the key change here is that this now refers to esql's Equals, not the ql base Equals.

Reply:
You could minimize the change by adding a templating method in the basic QL rule and overriding this in ESQL:
```
for (Expression exp : exps) {
   if (isEquals(exp)) {
     ....
   }
}

public boolean isEquals(Expression exp) {
        return exp instanceof QlEquals;
}

}

// rule in ESQL
public CombineDisjunctionsIn... extends ql.CombineDisjunctions {
      @Override
      protected boolean isEquals(Expression ex) {
             return ex instanceof EsqlEquals;
      }
}
```